core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemUpgradeRequest.java

```diff
@@ -0,0 +1,139 @@
+		}
+		
+		// or the selected extensionOf version is not present as a valid available upgrade
+		if (!availableUpgrades.contains(extensionOf)) {
+			throw new BadRequestException("Upgrades can only be performed to the next available version dependency.")
+				.withDeveloperMessage("Use '%s/<VERSION_ID>', where <VERSION_ID> is one of: '%s'", extensionOf.getCodeSystem(), availableUpgrades);
+		}
+		
+		// auto-generate the codeSystemId if not provided
+		// auto-generated upgrade IDs consist of the original CodeSystem's name and the new extensionOf dependency's URI
+		final String upgradeCodeSystemId;
+		if (codeSystemId == null) {
+			upgradeCodeSystemId = String.format("%s-%s-UPGRADE", codeSystem.getCodeSystem(), extensionOf.getPath() /*versionId*/); 
+		} else if (codeSystemId.isBlank()) {
+			throw new BadRequestException("'codeSystemId' property should not be empty, if provided");
+		} else {
+			upgradeCodeSystemId = codeSystemId;
+		}
+		
+		// create the same branch name under the new extensionOf path
+		String parentBranch = context.service(ResourceURIPathResolver.class).resolve(context, List.of(extensionOf)).stream().findFirst().get();
+		
+		String upgradeBranch = RepositoryRequests.branching().prepareCreate()
+			.setParent(parentBranch)
+			.setName(codeSystem.getCodeSystem())
+			.build()
+			.execute(context);
+		
+		// merge branch content from the current code system to the new upgradeBranch
+		Merge merge = RepositoryRequests.merging().prepareCreate()
+			.setSource(currentCodeSystem.getBranchPath())
+			.setTarget(parentBranch)
+			.setSquash(false)
+			.build()
+			.execute(context);
+		if (merge.getStatus() == Merge.Status.COMPLETED) {
+			context.log().error("Failed to sync source CodeSystem content to upgrade CodeSystem. Error: {}. Conflicts: {}", merge.getApiError(), merge.getConflicts());
+			throw new BadRequestException("Upgrade can not be performed due to content synchronization errors, see error logs.");
+		}
+		
+		// and lastly create the actual CodeSystem so users will be able to browse, access and complete the upgrade
+		return CodeSystemRequests.prepareNewCodeSystem()
+					.setShortName(upgradeCodeSystemId)
+					.setBranchPath(upgradeBranch)
+					// copy shared properties from the original CodeSystem
+					.setAdditionalProperties(currentCodeSystem.getAdditionalProperties())
+					.setCitation(currentCodeSystem.getCitation())
+					.setExtensionOf(extensionOf)
+					.setUpgradeOf(codeSystem)
+					.setIconPath(currentCodeSystem.getIconPath())
+					.setLocales(currentCodeSystem.getLocales())
+					.setLink(currentCodeSystem.getOrganizationLink())
+					.setName(String.format("Upgrade of '%s' to '%s'", currentCodeSystem.getName(), extensionOf))
+					.setRepositoryId(context.id())
+					.setTerminologyId(currentCodeSystem.getTerminologyId())
+					.build(context.id(), Branch.MAIN_PATH, context.service(User.class).getUsername(), String.format("Start upgrade of '%s' to '%s'", codeSystem, extensionOf))
```

Comment:
The commit message is a bit misleading, this is the final step 😄

Reply:
Yes, the final step, but from the upgrade point of view, it is the start of the upgrade and the commit marks that point, that's why it has this commit message.